### PR TITLE
[IOTDB-1760] Support other aggregations in group by fill

### PR DESCRIPTION
**Pull request by CRZbulabula**
Refactor GroupByFillDataSet.java to GroupByFillWithoutValueFilterDataSet.java.

Group by fill will be processed according to time series.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java [hunk 1]
```diff
@@ -51,10 +51,10 @@
 
 public class GroupByWithoutValueFilterDataSet extends GroupByEngineDataSet {
 
-  private static final Logger logger =
+  protected static final Logger logger =
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java:54]: You should construct a new Logger using the Subclass in itself.

File: server/src/main/java/org/apache/iotdb/db/query/executor/QueryRouter.java [hunk 2]
```diff
@@ -252,15 +256,28 @@ protected FillQueryExecutor getFillExecutor(FillQueryPlan plan) {
...
-        groupByFillPlan.getFillType(),
-        context,
-        groupByFillPlan);
+    //    if (logger.isDebugEnabled()) {
+    //      logger.debug(
+    //        "paths:"
+    //          + groupByFillPlan.getPaths()
+    //          + " level:"
+    //          + Arrays.toString(groupByFillPlan.getLevels()));
+    //    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/executor/QueryRouter.java:265]:
```suggestion

```

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java [hunk 3]
```diff
@@ -65,6 +68,30 @@ public LinearFill(TSDataType dataType, long queryTime, long beforeRange, long af
     this.afterRange = afterRange;
   }
 
+  public LinearFill() {}
+
+  public void convertRange(long startTime, long endTime) {
+    if (beforeRange % GroupByEngineDataSet.MS_TO_MONTH == 0) {
+      Calendar calendar = Calendar.getInstance();
```
Comment (cornmonster) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/LinearFill.java:75]: Please use the new date and time API defined in java8. See org.apache.iotdb.db.qp.utils.DatetimeUtils for examples.

File: server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java [hunk 4]
```diff
@@ -77,6 +80,18 @@ void constructFilter() {
     timeFilter = FilterFactory.and(lowerBound, TimeFilter.ltEq(queryTime));
   }
 
+  public void convertRange(long startTime) {
+    if (beforeRange % GroupByEngineDataSet.MS_TO_MONTH == 0) {
+      Calendar calendar = Calendar.getInstance();
```
Comment (cornmonster) [inline server/src/main/java/org/apache/iotdb/db/query/executor/fill/PreviousFill.java:85]: Please use the new date and time API defined in java8. See org.apache.iotdb.db.qp.utils.DatetimeUtils for examples.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java [hunk 5]
```diff
@@ -66,6 +66,83 @@ public GroupByEngineDataSet(QueryContext context, GroupByTimePlan groupByTimePla
...
+      // ie. startTIme = 1/31, interval = 1mo, curEndTime will be set to 2/29
+      retEndTime = Math.min(calcIntervalByMonth(interval), endTime);
+    } else {
+      retEndTime = Math.min(startTime + interval, endTime);
+    }
+    return new Pair<>(startTime, retEndTime);
+  }
+
+  protected Pair<Long, Long> getLastTimeRange() {
+    long retStartTime, retEndTime;
```
Comment (cornmonster) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java:82]: Coding style issue. Please do not define multiple variables in a single line.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java (hunk 1 above)
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithoutValueFilterDataSet.java]: done.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java (hunk 5 above)
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java]: done.

File: server/src/main/java/org/apache/iotdb/db/query/executor/QueryRouter.java (hunk 2 above)
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/query/executor/QueryRouter.java]: done

**Pull request merged**